### PR TITLE
ARROW-11310: [Rust] implement JSON writer

### PR DESCRIPTION
**Pull request by houqp**
Other than basic type support, this PR also covers the following cases:

- [x] List type
- [x] nested struct
- [x] nested list
- [x] list of struct
- [x] struct with list field

### CODE REVIEW

File: rust/arrow/src/json/writer.rs [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+    row_count: usize,
+    array: &ArrayRef,
+    col_name: &str,
+) {
+    let primitive_arr = as_primitive_array::<T>(array);
+    for (i, row) in rows.iter_mut().enumerate().take(row_count) {
+        row.insert(
+            col_name.to_string(),
+            primitive_arr
+                .value(i)
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:86]: This won't take nulls into account.

One way to go here is to use 

```rust
rows.iter_mut().zip(primitive_arr.iter()).take(row_count).for_each(|(row, maybe_value)| {
...
})
```

`maybe_value` will be of type `Option<T::Native>`, where `None` represents a null value.

File: rust/arrow/src/json/writer.rs [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    match array.data_type() {
+        DataType::Null => {
+            for row in rows.iter_mut().take(row_count) {
+                row.insert(col_name.to_string(), Value::Null);
+            }
+        }
+        DataType::Boolean => {
+            let arr = as_boolean_array(array);
+            for (i, row) in rows.iter_mut().take(row_count).enumerate() {
+                row.insert(col_name.to_string(), arr.value(i).into());
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:108]: Same here. Try using `arr.iter()`

File: rust/arrow/src/json/writer.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+        DataType::Float32 => {
+            set_column_by_primitive_type::<Float32Type>(rows, row_count, array, col_name)
+        }
+        DataType::Float64 => {
+            set_column_by_primitive_type::<Float64Type>(rows, row_count, array, col_name)
+        }
+        DataType::Utf8 => {
+            let strarr = as_string_array(array);
+            for (i, row) in rows.iter_mut().take(row_count).enumerate() {
+                row.insert(col_name.to_string(), strarr.value(i).into());
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:144]: same here.

File: rust/arrow/src/json/writer.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    use super::*;
+
+    #[test]
+    fn write_simple_rows() {
+        let schema = Schema::new(vec![
+            Field::new("a", DataType::Int32, false),
+            Field::new("b", DataType::Utf8, false),
+        ]);
+
+        let a = Int32Array::from(vec![1, 2, 3, 4, 5]);
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:248]: could have use `vec![Some(1), None, Some(3), Some(4,) Some(5)]`, so that we also test null values?

File: rust/arrow/src/json/writer.rs [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+            assert_eq!(
+                serde_json::from_str::<Value>(r).unwrap(),
+                serde_json::from_str::<Value>(e).unwrap()
+            );
+        }
+    }
+
+    #[test]
+    fn write_basic_rows() {
+        test_write_for_file("test/data/basic.json");
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:299]: AFAI understand this will populate a file and not delete it at the end.

Would it be possible to write to a byte stream or something instead of a file? Alternatively, use `tmp` or something?

**Review by jorgecarleitao — COMMENTED** [review]
Gave this one a quick look, and it looks good so far. I think the nulls are not accounted for.

I left some comments accordingly.

File: rust/arrow/src/json/writer.rs (hunk 1 above)
Comment (houqp) [reply rust/arrow/src/json/writer.rs:86]: Good catch, will update the code and add test cases.

File: rust/arrow/src/json/writer.rs (hunk 5 above)
Comment (houqp) [reply rust/arrow/src/json/writer.rs:299]: This function reads json test file from existing test data, serializes it to an in memory byte stream, then use that result to compare against the same test file from disk for diff. So there should be no disk write involved.

File: rust/arrow/src/array/cast.rs [hunk 6]
```diff
@@ -40,12 +40,20 @@ where
         .expect("Unable to downcast to dictionary array")
 }
 
-pub fn as_list_array<S: OffsetSizeTrait>(arr: &ArrayRef) -> &GenericListArray<S> {
+pub fn as_generic_list_array<S: OffsetSizeTrait>(arr: &ArrayRef) -> &GenericListArray<S> {
```
Comment (houqp) [inline rust/arrow/src/array/cast.rs:43]: breaking change to make naming more consistent with array type.

**Comment by houqp** [comment]
@jorgecarleitao @nevi-me ready for review.

File: rust/arrow/src/array/cast.rs [hunk 7]
```diff
@@ -40,12 +40,20 @@ where
...
     arr.as_any()
         .downcast_ref::<GenericListArray<S>>()
         .expect("Unable to downcast to list array")
 }
 
+pub fn as_list_array(arr: &ArrayRef) -> &ListArray {
```
Comment (jorgecarleitao) [inline rust/arrow/src/array/cast.rs:49]: Not sure whether we need an extra public function for a one-liner.

File: rust/arrow/src/json/writer.rs [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+        DataType::Int16 => primitive_array_to_json::<Int16Type>(array),
+        DataType::Int32 => primitive_array_to_json::<Int32Type>(array),
+        DataType::Int64 => primitive_array_to_json::<Int64Type>(array),
+        DataType::UInt8 => primitive_array_to_json::<UInt8Type>(array),
+        DataType::UInt16 => primitive_array_to_json::<UInt16Type>(array),
+        DataType::UInt32 => primitive_array_to_json::<UInt32Type>(array),
+        DataType::UInt64 => primitive_array_to_json::<UInt64Type>(array),
+        DataType::Float32 => primitive_array_to_json::<Float32Type>(array),
+        DataType::Float64 => primitive_array_to_json::<Float64Type>(array),
+        DataType::List(_) => as_list_array(array)
```
Comment (jorgecarleitao) [inline rust/arrow/src/json/writer.rs:139]:
```rust
        DataType::List(_) => as_large_list_array(array)
            .iter()
            .map(|maybe_value| match maybe_value {
                Some(v) => Value::Array(array_to_json_array(&v)),
                None => Value::Null,
            })
            .collect(),
```

for completeness?

**Review by jorgecarleitao — APPROVED** [review]
Great work, @houqp ! This looks super clean and IMO ready to go.

I would add a note to the readme where we discuss json readers. IMO this is a major feature!

**Comment by jorgecarleitao** [comment]
cc @andygrove @alamb @nevi-me @sunchao

File: rust/arrow/src/json/writer.rs [hunk 9]
```diff
@@ -0,0 +1,635 @@
...
+                Some(v) => Value::Array(array_to_json_array(&v)),
+                None => Value::Null,
+            })
+            .collect(),
+        DataType::Struct(_) => {
+            let jsonmaps =
+                struct_array_to_jsonmap_array(as_struct_array(array), array.len());
+            jsonmaps.into_iter().map(Value::Object).collect()
+        }
+        _ => {
```
Comment (nevi-me) [inline rust/arrow/src/json/writer.rs:151]: I think there'd be demand for dictionary support, but we can add that separately

**Review by nevi-me — APPROVED** [review]
Thanks for the implementation @houqp, minor comments

File: rust/arrow/src/json/writer.rs (hunk 9 above)
Comment (alamb) [reply rust/arrow/src/json/writer.rs:151]: Yes I agree -- on both counts :)

**Review by alamb — COMMENTED** [review]
I took a quick skim at this PR and it looks nice to me. Thank you @houqp

File: rust/arrow/src/array/cast.rs (hunk 7 above)
Comment (houqp) [reply rust/arrow/src/array/cast.rs:52]: @jorgecarleitao the main reason for adding that is to keep these two cast function names consistent with others. i.e. for a particular `FooBarArray` type, there exists a `as_foo_bar_array` cast function. However, if you have strong opinion on this, I am happy to remove it.

I have added the inline annotation.

File: rust/arrow/src/json/writer.rs (hunk 9 above)
Comment (houqp) [reply rust/arrow/src/json/writer.rs:158]: yeah, i was planning to add dictionary support, but this patch set already grew larger than I expected, so I decided to leave that to a follow up PR for easier review.

File: rust/arrow/src/array/cast.rs (hunk 7 above)
Comment (jorgecarleitao) [reply rust/arrow/src/array/cast.rs:52]: Makes sense. Let's keep it 👍

**Comment by alamb** [comment]
I plan to review this PR later today and merge unless I hear otherwise

**Review by alamb — APPROVED** [review]
Really nice work @houqp
